### ISSUE TITLE
ndk-glue: impossible to poll events in a non-blocking fashion

### ISSUE THREAD

**Issue by nwessing** [rust-windowing/android-ndk-rs#111]
it's impossible to poll events in ndk-glue without blocking the thread. In the NDK samples I have seen ALooper is used to poll events with a timeout value that can be configured to block or not. Where the ndk-glue module will read the file descriptor directly with a blocking call. This makes it impossible to program games, or any application that wants to keep running without  having to block for events/input.

**Comment by MarijnS95** [comment]
@nwessing When used with `winit` you have the option to set `ControlFlow` to `Poll` instead of `Wait`, this will disable blocking on the looper that `ndk-glue` sets up before calling `main()` (retrieved in at https://github.com/rust-windowing/winit/blob/d1a7749df56656e356840112ebcbb3c9f34de248/src/platform_impl/android/mod.rs#L75).

If you are not using `winit` you can still do the same, call `ThreadLooper::for_thread()` within `main()` to get the looper set up by `ndk-glue` - the one that contains the fd for the events - the same that is used for blocking rads in `poll_events`. From there you can call `poll_once_timeout` functions on `ThreadLooper` with a timeout of `0`.

**Comment by MarijnS95** [comment]
Related side-note:

@dvc94ch You added the event loop logic to `winit`, and it seems it's erroneously using the `data` field - which should be a pointer to user-data - to distinguish the event pipe from the `INPUT_QUEUE`:

https://github.com/rust-windowing/winit/blob/d1a7749df56656e356840112ebcbb3c9f34de248/src/platform_impl/android/mod.rs#L33-L37

Instead `ident` should be used there which is also set to `0` for the event pipe and `1` for the `INPUT_QUEUE`.

Shall I clean that up and send some PRs? Important is to get `winit` updated first, so that we can properly set this field to `std::ptr::null()` in `ndk-glue` after.

**Comment by nwessing** [comment]
I see, I was using the glue module directly, I am experimenting with using the Oculus Mobile SDK in Rust, and it doesn't make too much sense to use winit since I need to hook into a lot of android specific functionality directly anyways. I will use the winit/android implementation as guide then for how to use the ndk-glue crate. Thanks for pointing me in the right direction!

**Comment by dvc94ch** [comment]
@MarijnS95 ups, it is always 0 because it was set to `std::ptr::null()`.

**Comment by MarijnS95** [comment]
@dvc94ch it's actually set to 1 for the input queue which is why it works. Have a commit written up, will submit that in a bit ;)

**Comment by MarijnS95** [comment]
For reference, this is taken care of in https://github.com/rust-windowing/winit/pull/1826.

**Comment by kanerogers** [comment]
Chiming in to say that I'm also mucking about with the Oculus Mobile SDK in a similar manner to @nwessing, and some documentation on an easy, non-blocking way to get these events would be marvelous. 😄

**Comment by MarijnS95** [comment]
@kanerogers Does my comment above and the docstrings introduced in https://github.com/rust-windowing/android-ndk-rs/pull/112 provide that? Let us know if anything is missing, then I guess someone has to work that into a minimal looper example.

**Comment by nwessing** [comment]
@kanerogers I have this working pretty well at this point. I've defined the following function: 
```rust
pub const LOOPER_ID_MAIN: u32 = 0;
pub const LOOPER_ID_INPUT: u32 = 1;

pub fn poll_all_ms(block: bool) -> Option<ndk_glue::Event> {
    let looper = ThreadLooper::for_thread().unwrap();
    let result = if block {
        let result = looper.poll_all();
        result
    } else {
        looper.poll_all_timeout(std::time::Duration::from_millis(0u64))
    };

    match result {
        Ok(Poll::Event { ident, .. }) => {
            let ident = ident as u32;
            if ident == LOOPER_ID_MAIN {
                ndk_glue::poll_events()
            } else if ident == LOOPER_ID_INPUT {
                if let Some(input_queue) = ndk_glue::input_queue().as_ref() {
                    while let Some(event) = input_queue.get_event() {
                        if let Some(event) = input_queue.pre_dispatch(event) {
                            input_queue.finish_event(event, false);
                        }
                    }
                }

                None
            } else {
                unreachable!("Unrecognized looper identifer");
            }
        }
        _ => None,
    }
}
```

And it is called in my game loop like this: 
```rust
while !app.destroy_requested { // Main game loop
        loop { // event pump loop
            let block = !app.destroy_requested && app.ovr.is_none();
            if let Some(event) = poll_all_ms(block) {
                trace!("event: {:?}", event);
                app.handle_lifecycle_event(&event);
                app.update_vr_mode();
            } else {
                break;
            }
        }    
     // update and render
}
```
`app.ovr` is an `Option<*mut ovr_sys::ovrMobile>` (which is mutated in `update_vr_mode`) that is set to None when the window is destroyed or the app is in a paused state. (It follows the logic from the `VrCubeWorld_NativeActivity` Sample.

**Comment by kanerogers** [comment]
@MarijnS95 your documentation definitely helped! Thanks so much! ❤️ 

@nwessing We are doing the exact same thing at the same time. Thanks so much for this! 😄

**Comment by MarijnS95** [comment]
@kanerogers Neat! I'm hoping to get some remaining PRs reviewed/merged before merging #112 and bumping the minor version this weekend :grin:. Will try to stuff in a minimal looper example similar to the code pasted above :tada:!

**Comment by kanerogers** [comment]
Just chiming in to say this is working perfectly for me! :)
